### PR TITLE
FIX - Issue #408 - Column[ZonedDateTime] loses nanoseconds precision

### LINKED ISSUE #325: Default Column[Instant] instance loose nano seconds precision
### Anorm Version
- 2.6.8

### Operating System
- MacOS 11.1

### JDK

tested on both:
- OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_272-b10)
- OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.9+11)

### Expected Behavior

Please describe the expected behavior of the issue, starting from the first action.

1. insertion of a **java.time.Instant** value with nano seconds precision in a database table, using the standard **instantToStatement: ToStatement[Instant]** (https://github.com/playframework/anorm/blob/master/core/src/main/scala/anorm/ToStatementMisc.scala#L727)
2. read the inserted row, using the **columnToInstant: Column[Instant]** to build a **RowParser[Instant]** (https://github.com/playframework/anorm/blob/master/core/src/main/scala/anorm/Column.scala#L713)
3. I expect the value to be exactly the same as the value inserted step 1

### Actual Behavior

Please provide a description of what actually happens, working from the same starting point.

1. insert the row
2. read the row
3. the value read from the database is truncated to miliseconds, loosing the nano seconds precision. This is not what I expected.

### Reproducible Test Case

Here a complete and minimal example that illustrate the problem.

https://github.com/FXHibon/play-anorm-nano-precision

The failings test can be accessed here => https://github.com/FXHibon/play-anorm-nano-precision/blob/master/test/repositories/InstantRepositoryTest.scala#L39

### What else?

I noticed this problem when I switched my app from JDK 8 to JDK 11
Using JDK 8, **Instant** instances generated with `java.time.Instant.now` are **not** nano second precision
Using JDK 11, same code will generate instances with nano seconds precision
See screen below to illustrate this
image

Consequently, the reported issue is less likely to happen using JDK 8

I will quickly provide a pull request to fix the default **Column[Instant]**, so the nano seconds are not lost during the reading of sql result

### LINKED ISSUE #408: Default Column[ZonedDateTime] instance lose nanoseconds precision
### Anorm Version

- 2.6.10

### Operating System

- MacOS 11.1

### JDK

- OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.9+11)

### Library Dependencies

- PostgreSQL 9.6.12
- Driver "org.postgresql:postgresql:42.3.1"

### Expected Behavior

1. Insertion of a `java.time.ZonedDateTime` value with nano seconds precision in a database table, using the standard `zonedDateTimeToStatement: ToStatement[ZonedDateTime]` (https://github.com/playframework/anorm/blob/master/core/src/main/scala/anorm/ToStatementMisc.scala#L779)
2. Read the inserted row, using the `columnToZonedDateTime: Column[ZonedDateTime]` to build a RowParser[ZonedDateTime] (https://github.com/playframework/anorm/blob/master/core/src/main/scala/anorm/Column.scala#L814)
3. The returned value should be the same as the value inserted step 1

### Actual Behavior

1. Insertion of a `java.time.ZonedDateTime` value with nano seconds precision
2. Read the inserted row
3. The returned value is truncated to miliseconds, loosing the nano seconds precision

This issue is very similar to the issue #325 

### Reproducible Test Case

https://github.com/rrramiro/play-anorm-nano-precision

### PR DESCRIPTION
**Pull request by rrramiro**
# Pull Request Checklist

* [X] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [X] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [X] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [X] Have you [squashed your commits](https://www.playframework.com/documentation/2.4.x/WorkingWithGit#Squashing-commits)?
* [X] Have you added copyright headers to new files?
* [X] Have you checked that both Scala and Java APIs are updated?
* [X] Have you updated the documentation for both Scala and Java sections?
* [X] Have you added tests for any changed functionality?

## Fixes

Fixes #408

## Purpose

Fix the default instance of `Column[ZonedDateTime]` in the `JavaTimeColumn` trait, so it doesn't lose the nanosecond precision during the SQL result reads.

## Background Context

When reading an SQL result, `TIMESTAMP[TZ]` from database is mapped on the `java.sql.Timestamp` type. This type extends from `java.util.Date` but has a nanosecond precision. In current implementation the conversion is handle by `temporalValueTo` function:
```scala
  private def temporalValueTo[T](epoch: Long => T, description: String)(value: Any, meta: MetaDataItem): Either[SqlRequestError, T] = {
    val MetaDataItem(qualified, _, _) = meta
    value match {
      case date: java.util.Date => Right(epoch(date.getTime)) // Timestamp conversion
      case time: Long => Right(epoch(time))
      case TimestampWrapper1(ts) => Ts(ts)(t => epoch(t.getTime))
      case TimestampWrapper2(ts) => Ts(ts)(t => epoch(t.getTime))
      case _ => Left(TypeDoesNotMatch(s"Cannot convert $value: ${className(value)} to $description for column $qualified"))
    }
  }
```
`java.sql.Timestamp` is matched as a `java.util.Date` and `date.getTime` only provides milliseconds.

There was already a fix for the `Instant` type:

```scala
  implicit val columnToInstant: Column[Instant] = nonNull { (value, meta) =>
    val MetaDataItem(qualified, _, _) = meta

    value match {
      case date: LocalDateTime => Right(date.toInstant(ZoneOffset.UTC))
      case ts: java.sql.Timestamp => Ts(ts)(_.toInstant) //Timestamp conversion with nanoseconds
      case date: java.util.Date => Right(Instant ofEpochMilli date.getTime)
      case time: Long => Right(Instant ofEpochMilli time)
      case TimestampWrapper1(ts) => Ts(ts)(_.toInstant)
      case TimestampWrapper2(ts) => Ts(ts)(_.toInstant)
      case _ => Left(TypeDoesNotMatch(s"Cannot convert $value: ${className(value)} to Java8 Instant for column $qualified"))
    }
  }
```

My suggestion is to reuse this fix and to use `Instant` as the intermediate type for date/time conversion instead of `Long`.
So that all conversion will have to provide a function `Instant => T` when it was `Long => T`.

## References

`Column[Instant]`  fix  #325 and PR #326

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by rrramiro** [comment]
Thanks for the review @cchantep. You're right no need to use `Instant` for `LocalDateTime`

File: core/src/test/scala/anorm/JavaTimeSpec.scala [hunk 1]
```diff
@@ -188,6 +188,15 @@ class JavaTimeColumnSpec extends Specification {
           aka("parsed zoned date/time") must_=== date
       }
 
+    "be parsed from timestamp with nano precision" in {
+      val instantWithNanoPrecision = Instant.parse("2021-01-06T08:45:26.441477Z")
```
Comment (markarasev) [inline core/src/test/scala/anorm/JavaTimeSpec.scala:192]: Is it me or the tested instant has a millisecond precision instead of nanosecond?

Comment (rrramiro) [reply core/src/test/scala/anorm/JavaTimeSpec.scala:192]: The tested instant has `441477` **micro**seconds. The issue was happening for all the presisions under millisecond. So the test is not wrong, just the naming.

**Pull request merged**